File: python/socketd/socketd/transport/client/ClientConfig.py
from urllib.parse import urlparse

from socketd.exception.SocketDExecption import SocketDException
from socketd.transport.core.impl.ConfigBase import ConfigBase


class ClientConfig(ConfigBase):
    def __init__(self, url: str):
        super().__init__(True)

        idx = url.index("://")
        if idx < 2:
            raise SocketDException(f"The serverUrl invalid: {url}")

        self.__schema = url[:idx]

        if url.startswith("sd:"):
            url = url[3:]

        uri = urlparse(url)

        self.__linkUrl = "sd:" + url
        self.__url = url
        self.__host = uri.hostname
        self.__port = uri.port
        self.__schemaCleaned = uri.scheme

        if not self.__port:
            self.__port = 8602

        self.__connectTimeout = 10_000
        self.__heartbeatInterval = 20_000

        self.__autoReconnect = True

        self.__metaMap: dict[str, str] = dict()

    def get_link_url(self) -> str:
        return self.__linkUrl

    def get_url(self) -> str:
        return self.__url

    def get_schema(self) -> str:
        return self.__schema

    def get_host(self) -> str:
        return self.__host

    def get_port(self) -> int:
        return self.__port

    def get_meta_map(self) -> dict[str, str]:
        return self.__metaMap

    def meta_put(self, name, val):
        self.__metaMap[name] = val
        return self

    def get_heartbeat_interval(self) -> int:
        return self.__heartbeatInterval

    def heartbeat_interval(self, heartbeatInterval: int):
        self.__heartbeatInterval = heartbeatInterval
        return self

    def get_connect_timeout(self):
        return self.__connectTimeout

    def connect_timeout(self, connectTimeout):
        self.__connectTimeout = connectTimeout
        return self

    def is_auto_reconnect(self):
        return self.__autoReconnect

    def auto_reconnect(self, autoReconnect):
        self.__autoReconnect = autoReconnect
        return self

    def __str__(self):
        return f"ClientConfig{{schema='{self.__schemaCleaned}', " \
               f"charset='{self._charset}', " \
               f"url='{self.__url}', " \
               f"ioThreads={self._ioThreads}, " \
               f"codecThreads={self._codecThreads}, " \
               f"exchangeThreads={self._exchangeThreads}, " \
               f"heartbeatInterval={self.__heartbeatInterval}, " \
               f"connectTimeout={self.__connectTimeout}, " \
               f"idleTimeout={self._idleTimeout}, " \
               f"requestTimeout={self._requestTimeout}, " \
               f"streamTimeout={self._streamTimeout}, " \
               f"readBufferSize={self._readBufferSize}, " \
               f"writeBufferSize={self._writeBufferSize}, " \
               f"autoReconnect={self.__autoReconnect}, " \
               f"maxUdpSize={self._maxUdpSize}}}"

File: javascript/socketd/src/socketd/transport/client/ClientConfig.ts
import {ConfigBase} from "../core/Config";
import {StrUtils} from "../../utils/StrUtils";

export class ClientConfig extends ConfigBase {
    //协议架构（tcp, ws, udp, ...）
    private readonly _schema: string;
    private readonly _schemaCleaned: string;

    //连接地址
    private readonly _linkUrl: string;
    private readonly _url: string;
    private readonly _host: string;
    private readonly _port: number;
    private _metaMap: Map<string, string> = new Map<string, string>();

    //心跳间隔（毫秒）
    private _heartbeatInterval: number;

    //连接越时（毫秒）
    private _connectTimeout: number;

    //是否自动重链
    private _autoReconnect: boolean;

    constructor(url: string) {
        super(true);

        const idx = url.indexOf("://");
        if (idx < 2) {
            throw new Error("The serverUrl invalid: " + url);
        }

        this._schema = url.substring(0, idx);

        //支持 sd: 开头的架构
        if (url.startsWith("sd:")) {
            url = url.substring(3);
        }

        this._url = url;
        this._linkUrl = "sd:" + url;

        let _uri = StrUtils.parseUri(url);

        this._host = _uri.host;
        this._port = parseInt(_uri.port);
        this._schemaCleaned = _uri.protocol;

        if (this._port < 0) {
            this._port = 8602;
        }

        this._connectTimeout = 10_000;
        this._heartbeatInterval = 20_000;

        this._autoReconnect = true;
    }


    /**
     * 获取协议架构（tcp, ws, udp）
     */
    getSchema(): string {
        return this._schema;
    }

    /**
     * 获取链接地址
     */
    getLinkUrl(): string {
        return this._linkUrl;
    }

    /**
     * 获取连接地址
     */
    getUrl(): string {
        return this._url;
    }

    /**
     * 获取连接主机
     */
    getHost(): string {
        return this._host;
    }

    /**
     * 获取连接端口
     */
    getPort(): number {
        return this._port;
    }

    /**
     * 获取连接元信息字典
     */
    getMetaMap(): Map<string, string> {
        return this._metaMap;
    }

    metaPut(name: string, val: string): this {
        this._metaMap.set(name, val);
        return this;
    }


    /**
     * 获取心跳间隔（单位毫秒）
     */
    getHeartbeatInterval(): number {
        return this._heartbeatInterval;
    }

    /**
     * 配置心跳间隔（单位毫秒）
     */
    heartbeatInterval(heartbeatInterval: number): this {
        this._heartbeatInterval = heartbeatInterval;
        return this;
    }

    /**
     * 获取连接超时（单位毫秒）
     */
    getConnectTimeout(): number {
        return this._connectTimeout;
    }

    /**
     * 配置连接超时（单位毫秒）
     */
    connectTimeout(connectTimeout: number): this {
        this._connectTimeout = connectTimeout;
        return this;
    }

    /**
     * 获取是否自动重链
     */
    isAutoReconnect(): boolean {
        return this._autoReconnect;
    }

    /**
     * 配置是否自动重链
     */
    autoReconnect(autoReconnect: boolean): this {
        this._autoReconnect = autoReconnect;
        return this;
    }

    idleTimeout(idleTimeout: number): this {
        if (this._autoReconnect == false) {
            //自动重链下，禁用 idleTimeout
            this._idleTimeout = (idleTimeout);
            return this;
        } else {
            this._idleTimeout = (0);
            return this;
        }
    }

    toString(): string {
        return "ClientConfig{" +
            "schema='" + this._schemaCleaned + '\'' +
            ", charset=" + this._charset +
            ", url='" + this._url + '\'' +
            ", ioThreads=" + this._ioThreads +
            ", codecThreads=" + this._codecThreads +
            ", exchangeThreads=" + this._exchangeThreads +
            ", heartbeatInterval=" + this._heartbeatInterval +
            ", connectTimeout=" + this._connectTimeout +
            ", idleTimeout=" + this._idleTimeout +
            ", requestTimeout=" + this._requestTimeout +
            ", readBufferSize=" + this._readBufferSize +
            ", writeBufferSize=" + this._writeBufferSize +
            ", autoReconnect=" + this._autoReconnect +
            ", maxUdpSize=" + this._maxUdpSize +
            '}';
    }
}